juju/juju
# Ensure cross model app proxy is removed with the offer

Hopefully the final fix for cross model relations issues observed on site.
When an offer is removed, any remote proxies for the consuming side (in the offering model) also need to be deleted. This wasn't always happening, partly because the removal depended on there being relations in play. The logic to remove the app proxies has been moved so it is always run even if there are no relations.

Because there could be orphaned remote proxies, an upgrade step is added to delete them.

Finally, to guard against the possibility that reference counting across models might get out of sync and block relation and unit removal (in a single model things are done inside a txn but that's not possible across models), if the txn slice fails to be applied the first time, cross model relations are more forcibly removed the second time.

## QA steps

I used the repro scenario from site, primarily this bundle deployed to a model called `ceph`:
```
series: bionic
applications:
  ceph-mon:
    charm: cs:ceph-mon-55
    num_units: 3
    options:
      default-rbd-features: 1
      expected-osd-count: 3
      permit-insecure-cmr: True
  ceph-osd:
    charm: cs:ceph-osd-310
    num_units: 3
    options:
      osd-devices: /srv/ceph-osd
      bluestore: False
relations:
- - ceph-mon:osd
  - ceph-osd:mon
```

Then set up the relations:
```
juju offer ceph-mon:client ceph-mon-client
juju offer ceph-mon:radosgw ceph-mon-radogw

juju add-model consume
juju deploy cs:ceph-radosgw-296 --series bionic
juju relate admin/ceph.ceph-mon-radogw ceph-radosgw:mon
juju deploy glance --series bionic
juju relate admin/ceph.ceph-mon-client glance
juju deploy cs:keystone --series bionic
juju deploy cs:percona-cluster --series bionic
juju config percona-cluster min-cluster-size=1
juju relate glance percona-cluster
juju relate keystone percona-cluster
juju relate keystone glance
juju deploy cs:nova-compute --series bionic
juju relate nova-compute:image-service glance:image-service
juju relate nova-compute:ceph ceph-mon-client
```

You can now repeatedly 
```
juju switch ceph
juju remove-offer ceph-mon-radogw --force -y
juju offer ceph-mon:radosgw ceph-mon-radogw
juju switch consume
juju relate admin/ceph.ceph-mon-radogw ceph-radosgw:mon
```

For each iteration, you van see the relation created, joined, departed, broken hooks have run:
```
juju show-status-log -m consume ceph-radosgw/0
juju show-status-log -m ceph ceph-mon/0
```

## Bug reference

https://bugs.launchpad.net/charm-ceph-mon/+bug/1940983
